### PR TITLE
[AutoDiff] Destroy unused pullback direct results.

### PR DESCRIPTION
**Pull request by dan-zheng**
In `PullbackEmitter::visitApplyInst`, destroy unused pullback direct results.
This is needed for VJPs extracted from `@differentiable` function callees,
where the `@differentiable` function's differentiation parameter indices are
a superset of the active `apply` parameter indices.

Resolves TF-928.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -6975,13 +6975,23 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
...
+    // Record remaining pullback direct results as temporary values for cleanup.
+    // Needed to destroy unused pullback results for VJPs extracted from
+    // `@differentiable` function callees, when the `@differentiable` function's
+    // differentiation parameter indices are a superset of the active `apply`
+    // parameter indices.
+    while (allResultsIt != allResults.end()) {
+      auto unusedPullbackDirectResult = *allResultsIt++;
+      if (unusedPullbackDirectResult->getType().isAddress())
+        continue;
+      recordTemporary(unusedPullbackDirectResult);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: These unused results are not going to be used beyond this point. No need to record it as a temporary -- you can just `destroy_value` it.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks! Done in https://github.com/apple/swift/pull/28207/commits/a2af88f7f7653e735508076931fabc32bca195ac.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -6975,13 +6975,22 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
...
     }
+    // Destroy unused pullback direct results. Needed for pullback results from
+    // VJPs extracted from `@differentiable` function callees, where the
+    // `@differentiable` function's differentiation parameter indices are a
+    // superset of the active `apply` parameter indices.
+    while (allResultsIt != allResults.end()) {
+      auto unusedPullbackDirectResult = *allResultsIt++;
+      if (unusedPullbackDirectResult->getType().isAddress())
+        continue;
+      builder.emitDestroyOperation(loc, unusedPullbackDirectResult);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
      builder.emitDestroyValueOperation(loc, unusedPullbackDirectResult);
```
`SILBuilder::emitDestroyOperation` has a check and skips addresses, which you've already done above the call. So ideally you should call `SILBuilder::emitDestroyValueOperation` instead.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Okay! Done in https://github.com/apple/swift/pull/28207/commits/9783b0adf07f78ca98351909141099bca4be8c52.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**